Pattern matching (#213)
Kotlin/KEEP: proposals/pattern-matching.md

functionality through full blown pattern matching, which allows to type
check, equality test, and extract information in a single, intuitive
construct already in use in many popular languages.

A clear immediate advantage of this extension is avoiding nested `when`s, for
example.

The syntax proposed below aims to not introduce new keywords and leverage the
existing `when` idiom that the community has already grown used to, but
discussion is encouraged and welcome on how it can be improved. While this possible syntax is extensively discussed, the main focus of the proposal is not any specific pattern matching syntax, but rather the feature itself.

### Simple textbook example

```kotlin
data class Customer(val name: String, val age: Int, val email: String)
data class Prospect(val email: String, active: Boolean)
// ...

when(elem) {
  is Customer(name, age, addr) where { age >= 18 } ->
    Mail.send(addr, "Thanks for choosing us, $name!")

  is Prospect(addr, true) ->
    Mail.send(addr, "Please consider our product...")
}
```

The syntax proposed uses the already existing `is` construct to check the
type of the subject, but adds what semantically looks a lot like a

> destructuring delcaration with added equality checks. This approach is

Reviewer: ```suggestion
destructuring declaration with added equality checks. This approach is
```

 - It is much more verbose to write for nested patterns of more than one
 level, or where there is a lot of inspection of destructured elements.
 Consider how using `is` and `val` every time would look for the 
 [Arrow example](#arrow-example), which does not even make use of
more than 1 level of nesting nor guards:
  
```kotlin
fun <A> Kind<ForOption, A>.eqK(other: Kind<ForOption, A>, EQ: Eq<A>) =
    when(this.fix() to other.fix()) {
        is (is Some(val a), is Some(val b)) -> EQ.run { a.eqv(b) }
        is (is None, is None) -> true
        else -> false
    }

fun Ior<L, R>.eqv(other: Ior<L, R>): Boolean = when (this to other) {
    is (is Ior.Left(val a), is Ior.Left(val b)) -> EQL().run { a.eqv(b) }
    is (is Ior.Both(val al, val ar), is Ior.Both(val bl, val br)) ->
      EQL().run { al.eqv(bl) } && EQR().run { ar.eqv(br) }
    is (is Ior.Right(val a), is Ior.Right(val b)) -> EQR().run { a.eqv(b) }
    else -> false
}
```

#### Conclusion

There are two competing alternative syntax possibilites for pattern matching
in Kotlin. While this proposal argues for one that
- is slightly more restrictive (no nested `when` conditions)
- is less verbose
- encourages guards

> ...using `val` and `is` would make for a pattern mathcing that

Reviewer: ```suggestion
...using `val` and `is` would make for a pattern matching that
```

  - is more verbose (but offers some clarity)
  - allows inlining for equality checking
  - allows recursive `when` conditions (like `in 0..18`)

### Restricting matching to data classes only

> A possibilty suggested during the conception of this proposal was to restrict pattern matching to data classes. The argument would be to start with a samller size of 'matchable' elements in order to keep the inital proposal and feature as simple as possible, as it could be extended later down the line.

Reviewer: ```suggestion
A possibility suggested during the conception of this proposal was to restrict pattern matching to data classes. The argument would be to start with a smaller size of 'matchable' elements in order to keep the initial proposal and feature as simple as possible, as it could be extended later down the line.
```

This proposal argues **against** this restriction. Matching anything that implements `componentN()` has the important benefit of being able to match on 3rd party classes or interfaces that are not data classes, and to extend them for the sole purpose of matching. A notable example is `Map.Entry`, which is a Java interface.

## Limitations

### Matching on collections

An idiom in Haskell or Scala is to pattern match on collections. This relies
on the matched pattern 'changing' depending on the state of the collection.
Because this proposal aims to use `componentN()` for destructuring, such a
thing would not be possible in Kotlin as `componentN()` returns the Nth
element of the collection (instead of its tail for some `componentN()`).

This limitation is due to the fact that in Haskell, a list is represented
more similarly to how sealed classes work in Kotlin (and we can match on
those).

> **Pattern mathcing on collections is not the aim of this proposal**, but such

Reviewer: ```suggestion
**Pattern matching on collections is not the aim of this proposal**, but such
```
